### ISSUE TITLE
[Multiple Questions] Metadata Fetching, Libraries, and Other.

### ISSUE THREAD

**Issue by Z3ROGRAV1TYx** [happypandax/happypandax#135]
I hope I'm not too much of a pain for doing this, but I have a handful of questions about how stuff works and how to fix my issues after importing my library.

**1. First I'm a bit confused as to why it's fetching some things and not others. And why some things aren't being fetched completely?**

_For example: https://files.catbox.moe/6uhrww.png (nsfw warning)
In that photo, it seems to recognize it belongs to the Comiket 95 collection, and finds others in the series. It found the circle, and series name as well as artist.. but it is not grabbing photo art? (Same for those other ones below in the 'series').

It did find cover art for some others in that collection, but not all._ 

**2. Absolutely no tags seem to be generating.** 

_No tags seem to be getting automatically fetched at all. (Even on stuff where the rest of the metadata is found no issues. So it's not exactly an issue relating to my question 1.)

Unless I'm under the wrong impression that tags should be auto generated, it seems like they can only be manually inputted. Where is it grabbing metadata from? I would've thought it could be grabbing from EHentai and EXHentai, which is where a lot of my stuff is grabbed from, and I know for a fact that 99% of the stuff on there is tagged. So I don't see why it wouldn't be grabbing it._

**3. What exactly is the difference between Library, Gallery and Inbox?**

_Inbox is confusing as nothing is there, as for gallery and library, what would be the point of adding stuff to my library when it could all be viewed in the gallery? Is this mainly if someone is using multiple users, you have a gallery, but not all users want the same books, so they add what they want to the gallery? Kinda like a library? Or is there something else I am missing here?_

**4. Is there any way to "search" for a Doujin? Or import a link? To attach proper metadata to it if it's not being found?**

_Is it possible to manually search for a title or give HPX a link for a specific doujin so that it can grab the metadata? I see you can manually insert and edit metadata, but that can be time consuming for larger libraries. If not are there plans for this in the future?_

**5. Are Hentai Manga supported? (Official releases non doujin)? Or is the metadata fetcher only for Doujinshi?**

_If not are there plans for this in the future?_

**6. Is there an option to have Japanese Titles translated to English?**

_Basically is there a way to get Japanese Titles translated to English/Romaji, without changing the actual file name?_

Thank you very much, I look forward to any answers that can be provided! Hope I wasn't a pain ♥

**Comment by GentleViking** [comment]
1. Here is the [documentation](https://happypandax.github.io/faq.html) for scanning using tags like {gallery}, {collection} and so on  Other than that I don't know exactly how the folder names are getting parsed.

2. You need to add [plugins](https://happypandax.github.io/usage.html#installing-plugins). There is a plugin for ehentai and exhentai respectively. Once you downloaded them, you add an external link to the gallery you want to parse tags for (must be exhentai/ehentai link of course) and the metadata should be loaded by clicking the query metadata button on the top left. 

3. There are only two separate categories: inbox and library. You can further filter your results by selecting either "Collection" or "Gallery" As far as I know the inbox is like a plan to read folder where you can put galleries you wish to read soon like bookmarks. The Library is just the standard place where all of your galleries or collections go.

4. Yes refer to the answer for number 2. As far as I know there is no automatic metadata parsing tool without manually adding the external link yourself.

6. Currently I don't think there is a way to translate japanese titles to english. Maybe a future plugin or baked in translator will be made, but for now it just grabs the folder name. When you grab the metadata there is a possibility that the japanese and english title will come with it, but it won't be displayed on the main page.

7. Haven't tried that out yet, I'm also a rookie when it comes to networking.

Hopefully I helped answer some of your questions :) I'm also learning how most of this works

**Comment by Z3ROGRAV1TYx** [comment]
@GentleViking 
Will check out the rest in a bit when I have time, but I did try installing the plugins.
Trying to setup the EHLogin currently, but when I go to about>plugins and I installed it, It installed then it opens another webpage saying to login, with teh HappyPandaX prompt. I type in my HappyPandaX login and click enter, then I get an error saying "[AuthMissingCredentials|412]: [AuthMissingCredentials|412]: [InvalidMessage|403]: JSON dict missing 'user' key". I tried restarting browser, clearing cache etc.

What I have found is, when I login to HPX (Main) then go to about>plguins>ehentai login>open page
then it takes me to that HPX Plugins login, if I just close that page, then click Open page again on the other page, it opens to the prompt where you type in E-Hentai/ExHentai login.. BUT the input fields are just greyed out with a buffering symbol that isn't going away.

 Idk how to fix this. Idk if it is breaking because I'm somehow bypassing that HPX Plugins page login when I need to login, and it's just not letting me input the EHentai login because I skipped the HPX Login. Idk.

**Comment by GentleViking** [comment]
Maybe try using a guest account to see if it circumvents that bug.

**Comment by Z3ROGRAV1TYx** [comment]
<s>I have guest accounts disabled currently. I will try and enable it temporarily to see.</s>

That seemed to work. Weird.

**Comment by Z3ROGRAV1TYx** [comment]
Now, how do I go about removing all the old files, so I can rescan? 
I just reran the scan, and accidentally now duplicated all the entries .-.
Do I manually have to trash every single one of these then rescan..? >.>

**Comment by GentleViking** [comment]
Yeah, right now I can't find a way to delete all entries within the web client. I did the quick and dirty method of deleting the database file in the happypandax directory. Restarting the server will create a new clean database for you. (Only do this if you want to delete ALL entries)

If you are using sql database:
panda/data/database.db

If you are using postgres database:
just delete the happypandax table from postgres webclient.
------------------------------------------------------------------
I don't know of a way to rescan already scanned folders without some weird stuff happening.

**Comment by Z3ROGRAV1TYx** [comment]
Not seeing a database.db file inside of the installfolder (C:\Users\xx\AppData\Local\HappyPanda X\data)

I see a happypanda.db. Not sure if that is it?
https://files.catbox.moe/3xqxz6.png

**Comment by GentleViking** [comment]
That should be it. I forgot what the exact name of the sql database file was but it should be that.

**Comment by Z3ROGRAV1TYx** [comment]
Yeah I'll try it. I'm currently using sql db. That was default. Curiously is there any con to swapping to postgres? Looking into it, is it just an open source alternative? Or do I have the wrong idea. I'm +1 for open source so yeah.

**Comment by GentleViking** [comment]
Postgres is faster overall as far as I know. If you plan on having a sizable number of entries I'd go with it. The speed improvement is not drastic when I tested it, but it might be different if you have more than a thousand entries. No cons other than setting up for the first time.

**Comment by Z3ROGRAV1TYx** [comment]
Due to me not using that from first install, what files do I have to delete to set it up with postgres?
Tried just deleting the db file but seems to be giving error. Should I just completely unintall HPX and Reinstall?

**Comment by GentleViking** [comment]
Here are my db [settings](https://gyazo.com/53ff56f45c9b59ab6a8ce3994c5e9b9b). You need to install postgres from their website first and follow the instructions given there. Then set your db settings like I did with your own password. It should work without having to delete the sql db file. You can see your happypandax db by using the pgAdmin application that came with the postgres install.

**Comment by twiddli** [comment]
Hello,

Thank you @GentleViking for answering some of the questions. I'll try to clarify some of @GentleViking's responses.

> 1: First I'm a bit confused as to why it's fetching some things and not others. And why some things aren't being fetched completely?

When you import a gallery into HPX some metadata are parsed and derived from the filename of the gallery. With the E-Hentai naming convention this metadata sometimes includes names of the artists, circles and more. This is why those fields are filled out upon import.
It seems like some of your galleries (from your screenshot) are failing to generate a thumbnail. This could sometimes happen due to HPX not finding the image files. A classic issue is when a gallery has its image files contained in a sub-folder in the archive or folder. You can fix this by toggling the option `Treat nested folder as being transparent` upon import.

> 2: Absolutely no tags seem to be generating.
> 4: Is there any way to "search" for a Doujin? Or import a link? To attach proper metadata to it if it's not being found?

Like mentioned previously, HPX is able to derive some tags from the filename of the gallery alone, which is why some fields are already set upon import.
Further tagging should be done manually or automatically through the metadata fetcher.
With the plugins enabled, the metadata fetcher is able to fetch metadata from various sites. If no external links has been set for a gallery, a search for the gallery is performed on several sites by the plugins to see if it can find a match to retrieve metadata from. This should work in most cases for galleries with the e-hentai naming convention. If not, and you think it's a bug, please create an issue in the plugins repo.

You can see how the fetching fared by checking the gallery log.

> 3: What exactly is the difference between Library, Gallery and Inbox?

@GentleViking almost got it right.
Although I'd allow people to use these different separations of galleries however they want, **Inbox** was intended to be used for galleries that were not properly tagged and should be kept out of normal operations until it became properly tagged.
As for the *plan to be read* folder that @GentleViking is talking about, there is already a feature in HPX where you can *Save a gallery for later* intended to be used for that purpose, although quite primitive atm.

> 5: Are Hentai Manga supported? (Official releases non doujin)? Or is the metadata fetcher only for Doujinshi?

I'm not quite sure what you mean by official Hentai Manga and what exactly to support. #94 is for regular manga, like Naruto and One Piece.  By the way, that issue is currently stalled since the thing I requested was not done, it would be great if you could help out there. 

> 6: Is there an option to have Japanese Titles translated to English?

Just for clarification, any gallery metadata change in HPX does not reflect on the gallery source.
Also, are you asking for Japanese titles to be automatically translated to English by HPX? Please create a new issue for this suggestion. 

> 7: When Exposing HappyPandaX publicily, do I have to type in my public ipv4 address of the computer it's running on?

If you're accessing from a different device, then yes. Although there's some ways to avoid this, it's out of scope for HPX. I suggest bookmarking the site to avoid having to type the IP everytime.


> Trying to setup the EHLogin currently, but when I go to about>plugins and I installed it, It installed then it opens another webpage saying to login, with teh HappyPandaX prompt. I type in my HappyPandaX login and click enter, then I get an error saying `[AuthMissingCredentials|412]: [AuthMissingCredentials|412]: [InvalidMessage|403]: JSON dict missing 'user' key`. I tried restarting browser, clearing cache etc, also tried a new browser (My primary browser is Firefox).

This is related to #123 

> I was going to try and reinstall the E-Hentai login plugin, but when I click remove on the client it pops up with another error saying: `"Server(100) [HappypandaError|100]: NotImplementedError()"`

Yes, removing a plugin is not implemented yet. I might do it for the next update.
Removing a plugin is just analogous to disabling the plugin and then deleting the plugin files.

>To be quite honest, the way files are imported are kinda frustrating and odd. I don't even see a place where it shows my scanned folders. For example, in something like Jellyfin/Plex/Emby you enter the folder location, then it saves that folder and you can go back and edit settings. I'm not even seeing an area like that.

I have never used any of those services so I'm not familiar with that system but it sounds interesting. If you can, then please create a detailed suggestion issue for this feature. I also agree that the import process could be improved.
It can only be improved faster with feedback from users, so those are very welcome.

**Comment by Z3ROGRAV1TYx** [comment]
I highly appreciate your response, thank you for your time! @twiddli

Haven't had time to mess with it anymore yet, I will probably come back sometime in next few days to give more detailed responses.. in the meantime though the one question I should clarify on:

 >7: When Exposing HappyPandaX publicily, do I have to type in my public ipv4 address of the computer it's running on?

>If you're accessing from a different device, then yes. Although there's some ways to avoid this, it's out of scope for HPX. I suggest bookmarking the site to avoid having to type the IP everytime.


Sorry I should have been more clear on my question here.. I am referring to the 'host_web' field in config. 
Would this be changed from 0.0.0.0 to the public ipv4 address of my computer?

**Comment by twiddli** [comment]
`host_web: 0.0.0.0` means that it'll serve on all your interfaces (localhost, ipv4 address of your computer, and other adressess that you might have). You can set `host_web: <ipv4 address>` if you want to limit it to only that specific `<ipv4 address>`
